design-documents/service-isolation/split-extensions.md

```diff
@@ -0,0 +1,81 @@
+The desired state that all modules/extensions should depend only on used component's parts like API, UI, etc. but not on whole component like now.
+
+![Components Dependencies](img/components-dependencies.png)
+
+## Split strategy
+
+There are multiple strategies how to split existing modules and extensions. Each strategy may imply a gradual separation into new modules.
+A tool might be used to split modules like in [prototype](https://github.com/magento-architects/magento2/tree/split-framework/app/code/Magento). It does not guarantee that all modules can be split automatically but minimize the routine work.
+
+### New Module - new class namespace
+
+This strategy assumes splitting existing modules into new with namespaces changes according to the new modules names.
+All dependant code should use new class namespaces.
+
+Desired state for dependencies:
+
+```json
+"require": {
+    "magento/module-catalog-api": "^104.0.0",
+    "magento/module-catalog-adminui": "^104.0.0",
+    "magento/module-catalog-ui": "^104.0.0"
+}
+```
+
+Pros:
+ - Introduce "right" dependencies
+ - Dependant code use only needed modules
+ 
+Cons:
+ - Cause BIC as all dependant code should you new namespaces
+ - All extensions should be affected
+ 
+### New module - old class namespace
+ 
+This approach proposes split modules but leave class namespaces as is. So all dependant code should require new packages in `composer.json` but does not require changes in codebase.
+ 
+ Desired state for dependencies:
+ 
+ ```json
+"require": {
+    "magento/module-catalog-api": "^104.0.0",
+    "magento/module-catalog-adminui": "^104.0.0",
+    "magento/module-catalog-ui": "^104.0.0"
+}
+```
+ 
+Pros:
+ - Introduce "right" dependencies
+ - Dependant code use only needed modules
+ - No changes in the dependant code
+  
+Cons:
+ - Dependant modules and extensions should update `composer.json`
+ - Extensions should split their modules in the same way as core modules and specify the correct dependencies
+
+### New module - old composer
+
+This strategy proposes to expose all new modules via one (currently existing) `composer.json` and much described in [this proposal](https://github.com/magento/architecture/issues/88).
+The main idea that other modules will user the same dependencies and namespaces.
+
+ ```json
+"require": {
+    "magento/module-catalog": "^103.0.0|^104.0.0"
+}
+```
+
+Pros:
+ - No changes in the dependant code
+ - Allows specify the 
+  
+Cons:
+ - `composer.json` anyway should be changed to specify needed version of module
+ - All modules and extensions still have coupling to all dependant code if ti does not use
+ - Extensions anyway should split their modules in the same way as core modules and specify the correct dependencies
+ - Such approach won't get benefits of modules decoupling as module|extension still depends on the whole component
```

Comment:
Maybe at some point we could stop including meta package that aggregates all extensions by default. But if some extensions have dependency on meta package, they would download it and continue to work. There are would be no benefits from modularity, but it would be backwards compatible.

Reply:
`Component` means for instance Catalog functionality, after splitting there will be multiple modules which aggregate the same component behavior

Reply:
What do you mean by catalog functionality? Modules that Catalog module will be separated on or all of the modules related to catalog (including Bundle, ConfigurableProduct, etc)?

We can also aggregate modules main module will be split on using meta package that will have main module name. If the new module keep old namespace, it would be backwards compatible. Could you add it also to the list?

Reply:
> What do you mean by catalog functionality? Modules that Catalog module will be separated on or all of the modules related to catalog (including Bundle, ConfigurableProduct, etc)?

Both, new modules from existing Catalog modules + Bundle, Configurable and all related as they represent catalog functionality.

> We can also aggregate modules main module will be split on using meta package that will have main module name. If the new module keep old namespace, it would be backwards compatible. Could you add it also to the list?

The last strategy describes this approach but based on composer packages, not a metapackage. I'm not sure about metapackage approach as 3rd party extensions depend on magento-module packages, not metapackages.